streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -38,346 +40,644 @@
+            // end offset is not initialized meaning that it is from a standby task (i.e. it should never end)
+            return false;
+        } else if (endOffset == 0) {
+            // this is a special case, meaning there's nothing to be restored since the changelog has no data
+            // of the changelog is a source topic and there's no committed offset
+            return true;
+        } else if (currentOffset == null) {
+            // current offset is not initialized meaning there's no checkpointed offset,
+            // we would start restoring from beginning and it does not end yet
+            return false;
         } else {
-            log.debug("Skip re-adding restorer for changelog {}", restorer.partition());
+            // note the end offset returned of the consumer is actually the last offset + 1
+            return currentOffset + 1 < endOffset;
         }
+    }
+
+    // Once some new tasks are created, we transit to restore them and pause on the existing standby tasks. It is
+    // possible that when newly created tasks are created the thread and its changelog reader are still restoring existing
+    // active tasks, and hence this function is idempotent and can be called multiple times.
+    //
+    // NOTE: even if the newly created tasks do not need any restoring, we still first transit to this state and then
+    // immediately transit back -- there's no overhead of transiting back and forth but simplifies the logic a lot.
+    // TODO K9113: this function should be called by stream thread
+    public void transitToRestoreActive() {
+        log.debug("Transiting to restore active tasks: {}", changelogs);
 
-        needsInitializing.add(restorer.partition());
+        // pause all partitions that are for standby tasks from the restore consumer
+        pauseChangelogsFromRestoreConsumer(standbyRestoringChangelogs());
+
+        state = ChangelogReaderState.ACTIVE_RESTORING;
     }
 
-    public Collection<TopicPartition> restore(final RestoringTasks active) {
-        if (!needsInitializing.isEmpty()) {
-            initialize(active);
+    // Only after we've completed restoring all active tasks we'll then move back to resume updating standby tasks.
+    // This function is NOT idempotent: if it is already in updating standby tasks mode, we should not call it again.
+    //
+    // NOTE: we do not clear completed active restoring changelogs or remove partitions from restore consumer either
+    // upon completing them but only pause the corresponding partitions; the changelog metadata / partitions would only
+    // be cleared when the corresponding task is being removed from the thread.
+    // TODO K9113: this function should be called by stream thread
+    public void transitToUpdateStandby() {
+        if (state != ChangelogReaderState.ACTIVE_RESTORING) {
+            throw new IllegalStateException("The changelog reader is not restoring actove tasks while trying to " +
```

Comment:
I will include the changelogs in the exception message.

Since this is a fatal error, the illegal state exception would never be swallowed and will always be printed to std. So I feel logging another error would have duplicated entries on log files.